src/access/AccessRestrictionProvider.cpp

```diff
@@ -0,0 +1,252 @@
+    {
+        if (l == &listener)
+        {
+            return;
+        }
+
+        if (l->mNext == nullptr)
+        {
+            l->mNext       = &listener;
+            listener.mNext = nullptr;
+            return;
+        }
+    }
+}
+
+void AccessRestrictionProvider::RemoveListener(Listener & listener)
+{
+    if (mListeners == &listener)
+    {
+        mListeners     = listener.mNext;
+        listener.mNext = nullptr;
+        return;
+    }
+
+    for (Listener * l = mListeners; l != nullptr; l = l->mNext)
+    {
+        if (l->mNext == &listener)
+        {
+            l->mNext       = listener.mNext;
+            listener.mNext = nullptr;
+            return;
+        }
+    }
+}
+
+CHIP_ERROR AccessRestrictionProvider::SetCommissioningEntries(const std::vector<Entry> & entries)
+{
+    // check that the input entries are valid
+    for (auto & entry : entries)
+    {
+        if (!IsEntryValid(entry))
+        {
+            ChipLogError(DataManagement, "AccessRestrictionProvider: invalid entry");
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    mCommissioningEntries = entries;
+
+    for (Listener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+    {
+        listener->CommissioningRestrictionListChanged();
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessRestrictionProvider::SetEntries(const FabricIndex fabricIndex, const std::vector<Entry> & entries)
+{
+    std::vector<Entry> updatedEntries;
+
+    // check that the input entries are valid
+    for (auto & entry : entries)
+    {
+        if (!IsEntryValid(entry))
+        {
+            ChipLogError(DataManagement, "AccessRestrictionProvider: invalid entry");
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+
+        Entry updatedEntry       = entry;
+        updatedEntry.fabricIndex = fabricIndex;
+        updatedEntries.push_back(updatedEntry);
+    }
+
+    mFabricEntries[fabricIndex] = std::move(updatedEntries);
+
+    for (Listener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+    {
+        listener->RestrictionListChanged(fabricIndex);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+bool AccessRestrictionProvider::StandardAccessRestrictionExceptionChecker::AreRestrictionsDisallowed(const SubjectDescriptor & subjectDescriptor,
+                                                                         const RequestPath & requestPath)
+{
+    if (requestPath.endpoint == 0 ||
+        requestPath.cluster == app::Clusters::NetworkCommissioning::Id ||
+        requestPath.cluster == app::Clusters::Descriptor::Id)
+    {
+        return true;
+    }
+
+    return false;
+}
+
+CHIP_ERROR AccessRestrictionProvider::CheckForCommissioning(const SubjectDescriptor & subjectDescriptor,
+                                                            const RequestPath & requestPath)
+{
+    if (mExceptionChecker.AreRestrictionsDisallowed(subjectDescriptor, requestPath))
```

Comment:
This MUST be moved in `Check()` and always apply, not just in commissioning. See latest pseudo code in spec PR.

Reply:
done